src/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptionProviderTests.cs

```diff
@@ -9,33 +9,361 @@ namespace System.ComponentModel.DataAnnotations.Tests
         {
             AssertExtensions.Throws<ArgumentNullException>("associatedMetadataType", () => new AssociatedMetadataTypeTypeDescriptionProvider(typeof(string), null));
         }
 
         [Fact]
-        public static void Validate_GetTypeDescriptor_ReturnsTypeDescriptor_ForTypeWithCustomDescription()
+        public void GetTypeDescriptor_MetadataHasFieldsNotPresentOnClass_ThrowsInvalidOperationException()
         {
-            var provider = new AssociatedMetadataTypeTypeDescriptionProvider(typeof(SomeClassWithMetadata));
-            ICustomTypeDescriptor typeDescriptor = provider.GetTypeDescriptor(typeof(SomeClassWithMetadata), null);
-            PropertyDescriptorCollection props = typeDescriptor.GetProperties();
-            PropertyDescriptor firstNameProp = props[nameof(SomeClassWithMetadata.FirstName)];
-            string displayName = firstNameProp.DisplayName;
+            var provider = new AssociatedMetadataTypeTypeDescriptionProvider(typeof(ClassWithInvalidMetadata));
+            Assert.Throws<InvalidOperationException>(() => provider.GetTypeDescriptor(typeof(ClassWithInvalidMetadata), null));
+        }
+
+        [Fact]
+        public void GetTypeDescriptorGetAttributes_NoAssociatedMetadataTypeWithoutMetadataTypeAttribute_ReturnsExpected()
+        {
+            void PerformTest(AssociatedMetadataTypeTypeDescriptionProvider provider)
+            {
+                ICustomTypeDescriptor typeDescriptor = provider.GetTypeDescriptor(typeof(ClassWithMetadata), null);
+                AttributeCollection attributes = typeDescriptor.GetAttributes();
+                Assert.Equal(1, attributes.Count);
+                Assert.Equal("typeName", Assert.IsType<TypeConverterAttribute>(attributes[typeof(TypeConverterAttribute)]).ConverterTypeName);
+            }
+
+            // Perform multiple times to test caching behaviour.
+            PerformTest(new AssociatedMetadataTypeTypeDescriptionProvider(typeof(ClassWithMetadata)));
+            PerformTest(new AssociatedMetadataTypeTypeDescriptionProvider(typeof(ClassWithMetadata)));
+        }
+
+        [Fact]
+        public void GetTypeDescriptorGetAttributes_NoAssociatedMetadataTypeWithMetadataTypeAttribute_ReturnsExpected()
+        {
+            void PerformTest(AssociatedMetadataTypeTypeDescriptionProvider provider)
+            {
+                ICustomTypeDescriptor typeDescriptor = provider.GetTypeDescriptor(typeof(ClassWithMetadataOnAnotherClass), null);
+                AttributeCollection attributes = typeDescriptor.GetAttributes();
+                Assert.Equal(2, attributes.Count);
+                Assert.Equal(typeof(ClassWithMetadata), Assert.IsType<MetadataTypeAttribute>(attributes[typeof(MetadataTypeAttribute)]).MetadataClassType);
+                Assert.Equal("typeName", Assert.IsType<TypeConverterAttribute>(attributes[typeof(TypeConverterAttribute)]).ConverterTypeName);
+            }
+
+            // Perform multiple times to test caching behaviour.
```

Comment:
What is the actual caching behavior being tested?  Should the test also be validating that the same objects are returned on subsequent calls?  That would be another reason not to use this local function approach.

Reply:
The cache is actually static, so not necessarily on the same instance. Updated the comments to clarify this